contributors/design-proposals/storage/dynamic_volume_limit.md

```diff
@@ -0,0 +1,108 @@
+# Dynamic attached volume limits
+
+## Goals
+
+Currently the number of volumes attachable to a node is either hardcoded or only configurable via an environment variable. Also
+existing limits only apply to well known volume types like EBS, GCE and is not available to all volume plugins.
+
+This proposal enables any volume plugin to specify those limits and also allows same volume type to have different volume
+limits depending on type of node.
+
+## Implementation Design
+
+### Prerequisite
+
+* This feature will be protected by an alpha feature gate, so as API and CLI changes needed for it. We are planning to call
+  the feature `DynamicVolumeThreshold`.
+
+### API Changes
+
+There is no API change needed for this feature. However existing `node.Status.Capacity` and `node.Status.Allocatable` will
+be extended to cover volume limits available on the node too.
+
+The key name that will store volume will be start with prefix `storage-limits-`. The volume limit key will respect
+format restrictions applied to Kubernetes Resource names. Volume limit key for existing plugins might look like:
+
+
+* `storage-limits-aws-ebs`
+* `storage-limits-gce-pd`
+* `storage-limits-csi-<driver_name>`
+
+`IsScalarResourceName` check will be extended to cover storage limits:
```

Comment:
Does this function ensure that users cannot request this resource normally in their Pod?

Reply:
This check in itself will not prevent users from requesting this inside pods. But since name does not adhere to StandardResourceName conventions, so users can not request it. I will update validations to be double sure that users can not request this resource.

Reply:
Added a note about, how are are making sure that this resource name can not be used in pods.

Reply:
How about NumberofAttachedVolumesLimit to make it more obvious?

Reply:
The `IsScalarResourceName` function is already used for checking whether a resource can be added to node's Capacity and Allocatable. We just have to extend the check to permit new volume limit values. We can't define new function without making bigger change in scheduler.

Reply:
i agree `storage-limits` is very generic, if this mainly limiting the number of volumes attached per node, we should call it so

Reply:
my point was what kind of storage limits is `storage-limits-aws-ebs` , is it size of volume that can be created, number of volumes attached to a node, or something else ?

Reply:
> Does this function ensure that users cannot request this resource normally in their Pod?

I had the same question. Are we already using capacity/allocatable resources reported by nodes to represent resources not requested/consumed by pods?

> since name does not adhere to StandardResourceName conventions, so users can not request it. I will update validations to be double sure that users can not request this resource.

we have to be careful not to box ourselves in here... if I'm reading this correctly, this PR is adding a new use of allocatable resources reported by nodes:
* known resources pods can request (cpu, memory, hugepages, etc)
* arbitrary resources pods can request (device drivers, etc)
* known resources pods cannot request (added by this PR?)

if using allocatable resources for scheduling/tracking things other than pods makes sense, does it seem likely we'll end up wanting to use arbitrary resources in that area as well? (thinking about other finite resources CSI impls would want to report on the node side and have affect PV scheduling). If so, how would those be namespaced/partitioned?

Reply:
Discussed this offline a bit and it is true that apart from this resource, there is no other resource that is available on the node but not directly requestable by the pod and it will be nice to segment resources that are available on the node but not requestable by the pods directly. 

However I think that, we can revisit that decision once we know more. cc @derekwaynecarr

Reply:
Max pod per node limit is a node capacity resource that cannot be explicitly requested by pods